helper/resource/testing.go

```diff
@@ -1632,6 +2023,51 @@ func TestCheckOutput(name, value string) TestCheckFunc {
 	}
 }
 
+// Deprecated: State checks using [TestStep.Check] have been superseded by [TestStep.ConfigStateChecks].
+// Use the built-in [statecheck.ExpectKnownOutputValue], and
+// [statecheck.ExpectKnownOutputValueAtPath] state checks, can be used in
+// combination with a [custom] [knownvalue.Check] instead.
+// TestMatchOutput function will be removed in the next major version.
+//
+// The following is an example of using [statecheck.ExpectKnownOutputValueAtPath]
+// in combination with a [custom] [knownvalue.Check] to replicate the behaviour of
+// TestMatchOutput.
+//
+//	package example_test
+//
+//	import (
+//		"testing"
+//
+//		"github.com/hashicorp/terraform-plugin-testing/helper/resource"
+//		"github.com/hashicorp/terraform-plugin-testing/statecheck"
+//		"github.com/hashicorp/terraform-plugin-testing/tfjsonpath"
+//	)
+//
+//	func TestExpectKnownOutputValueAtPath_CheckState_String_Custom(t *testing.T) {
+//		t.Parallel()
+//
+//		resource.Test(t, resource.TestCase{
+//			// Provider definition omitted.
+//			Steps: []resource.TestStep{
+//				{
+//					// Example resource containing a computed string attribute named "computed_attribute"
+//					Config: `resource "test_resource" "one" {}
+//
+//							output test_resource_one_output {
+//								value = test_resource.one
+//							}`,
+//					ConfigStateChecks: []statecheck.StateCheck{
+//						statecheck.ExpectKnownOutputValueAtPath(
+//							"test_resource_one_output",
+//							tfjsonpath.New("computed_attribute"),
+//							StringContains("str")),
```

Comment:
Similar here regarding some form of built-in string regular expression check.

Reply:
`knownvalue.StringRegularExpression` has been added and the docs updated accordingly.

Reply:
> A note (that we don't necessarily need to solve in this PR, since it's already pretty large)

`knownvalue.StringRegularExpression` won't be a replacement for certain `TestMatchResourceAttr` checks that assumed the testing framework would treat a state value that is `json.Number` as a string. Like this example on the time provider (where hour, minute, etc. are all number values in state)

https://github.com/hashicorp/terraform-provider-time/blob/87a1ba82545baa44dc9c6770c4d5342aa617c7f2/internal/provider/resource_time_static_test.go#L26-L29

This seems more to the edge in-terms of use-case, but thought I would mention it as I just came across it.

Reply:
My suggestion here is that it would be nice to support known value checks with the correct types and semantics rather than relying on certain implementation details that could change over time. We should encourage replacing workarounds like the one I had written there to ensure some level of full coverage. 😅 

More concretely, I would propose known value checks such as `Int64AtLeast(min int64)`, `Int64AtMost(max int64)`, `Int64Between(min, max int64)`, so this sort of checking would become `statecheck.ExpectKnownValue("time_static.test", tfjsonpath.New("day"), knownvalue.Int64Between(1, 31))`. 

There may also be a decent way to directly check via `time.Now()` as well by calling that in the test function and then passing its attribute values to checks, but I was worried about the slight time drift that will occur from when that function call occurs versus when the testing logic would actually plan/apply the resource.

Reply:
I replaced the awkward `NumberRegularExpression` with a more reasonable `Int64Between` in that linked `time` provider PR: https://github.com/hashicorp/terraform-provider-time/commit/d73f9a9168e7dd1554fe1fbe2cb8f4f676405852

If we end up moving that into `terraform-plugin-testing` we can replace it later 👍🏻